### DISCUSSION TITLE
partial_fit for incremental training of HT or EFDT in river

### DISCUSSION THREAD

**Discussion by Piyushbalwani** [online-ml/river#624] [General]
Hi Team,

As it is mentioned in [https://riverml.xyz/latest/user-guide/mini-batching/ ](https://riverml.xyz/latest/user-guide/mini-batching/), **`river`** package learn with one sample at a time. But in `skmultiflow`, there was implementation of partial_fit, which can fit multiple instances. In river , I found only **`learn_one`** for HT andEFDT. 
Is there any equivalent function or alternative way to fit multiple instances?

[https://scikit-multiflow.readthedocs.io/en/stable/api/generated/skmultiflow.trees.HoeffdingTreeClassifier.html](https://scikit-multiflow.readthedocs.io/en/stable/api/generated/skmultiflow.trees.HoeffdingTreeClassifier.html)

Regards,
Piyush

**Comment by smastelini** [comment]
Hi @Piyushbalwani. The partial_fit method in skmultiflow simply did a loop over each input example and called a private method that trained in an instance-by-instance regimen. So there were no real mini-batching/vectorization speed-ups.

It is in our plans to support real mini-batching capabilities in the trees in the future. But for now, the way to go is to iterate over instances and call `learn_one` for each instance.

**Reply by Piyushbalwani** [reply]
Thanks @smastelini. I was unaware of `skmultiflow partial_fit` instance-by-instance implementation. 
So, can we train the model with `skmultiflow` with initial large dataframe and save the model. Then load the model and do instance-by-instance incremental training?
Or is there any other workaround as dataframe is quite big?  

Regards,
Piyush

**Reply by smastelini** [reply]
The best alternative will depend on your usage. If your running everything locally you have a few options:

- `stream.iter_csv`, `stream.iter_arff`, or other utilities from the `stream` submodule if your data is saved in disk. Note that these iterators do not load the whole data into memory.
- `stream.iter_pandas` if your data frame is already loaded in memory. You can also use `iter_array` when using numpy objects :)

In all the cases, the tree models will learn instance by instance. You can perform some sort of warm-start with a subset of the examples, for sure. Nonetheless, `learn_one` is currently the only way to update the trees.

We hope to leverage vectorized operations in the future and provide `learn_many`/`predict_many` methods.
